File: Cake List/MasterViewController.m
//
//  MasterViewController.m
//  Cake List
//
//  Created by Stewart Hart on 19/05/2015.
//  Copyright (c) 2015 Stewart Hart. All rights reserved.
//

#import "MasterViewController.h"
#import "CakeCell.h"

@interface MasterViewController ()
@property (strong, nonatomic) NSArray *objects;
@end

@implementation MasterViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.refreshControl = [[UIRefreshControl alloc] init];
    self.refreshControl.backgroundColor = [UIColor orangeColor];
    self.refreshControl.tintColor = [UIColor whiteColor];
    [self.refreshControl addTarget:self
                            action:@selector(loadCakeData)
                  forControlEvents:UIControlEventValueChanged];
    
    [self loadCakeData];
}

#pragma mark - Table View
- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return self.objects.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    CakeCell *cell = (CakeCell*)[tableView dequeueReusableCellWithIdentifier:@"CakeCell"];
    
    NSDictionary *object = self.objects[indexPath.row];
    cell.titleLabel.text = object[@"title"];
    cell.descriptionLabel.text = object[@"desc"];

    NSURL *aURL = [NSURL URLWithString:object[@"image"]];
    
    //Asynchronously download cake image
    [self downloadCakeImage:aURL forImageView:cell.cakeImageView];
    
    return cell;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath{
    [tableView deselectRowAtIndexPath:indexPath animated:YES];
}

- (void)loadCakeData{
    
    NSURL *url = [NSURL URLWithString:@"https://gist.githubusercontent.com/hart88/198f29ec5114a3ec3460/raw/8dd19a88f9b8d24c23d9960f3300d0c917a4f07c/cake.json"];
    
    NSData *data = [NSData dataWithContentsOfURL:url];
    
    NSError *jsonError;
    id responseData = [NSJSONSerialization
                       JSONObjectWithData:data
                       options:kNilOptions
                       error:&jsonError];
    if (!jsonError){
        self.objects = responseData;
        [self.tableView reloadData];
        [self.refreshControl endRefreshing];
    } else {
        [self showErrorDialog];
    }
    
}

- (void)downloadCakeImage:(NSURL *)url forImageView:(UIImageView *)imageView
{
    imageView.alpha = 0;
    NSMutableURLRequest *request = [NSMutableURLRequest requestWithURL:url];
    [NSURLConnection sendAsynchronousRequest:request
    queue:[NSOperationQueue mainQueue]
    completionHandler:^(NSURLResponse *response, NSData *data, NSError *error) {
       if ( !error )
       {
           UIImage *image = [[UIImage alloc] initWithData:data];
           [UIView animateWithDuration:0.5 delay:0.0 options:UIViewAnimationOptionCurveEaseInOut animations:^{
               imageView.image = image;
               imageView.alpha = 1;
           } completion:^(BOOL finished) {}];
       } else {
           NSLog(@"%@", error);
       }
    }];
}

- (void)showErrorDialog
{
    UIAlertController * alert = [UIAlertController
                                 alertControllerWithTitle:@"Error"
                                 message:@"There was a problem loading the cakes."
                                 preferredStyle:UIAlertControllerStyleAlert];
    
    UIAlertAction* okButton = [UIAlertAction
                                actionWithTitle:@"OK"
                               style:UIAlertActionStyleDefault
                                handler:^(UIAlertAction * action) {}];
    
    UIAlertAction* retryButton = [UIAlertAction
                               actionWithTitle:@"Retry"
                               style:UIAlertActionStyleDefault
                               handler:^(UIAlertAction * action) {
                                   [self loadCakeData];
                               }];
    
    [alert addAction:okButton];
    [alert addAction:retryButton];
    
    [self presentViewController:alert animated:YES completion:nil];
}

@end

File: CakeList/CakeList/Controllers/MasterViewController.swift
//
//  MasterViewController.swift
//  CakeList
//
//  Created by Adam Bandeira Lynas on 15/04/2019.
//  Copyright © 2019 Limit Break Technology Ltd. All rights reserved.
//

import UIKit

class MasterViewController: UITableViewController {

    var objects = [CakeItem]()


    override func viewDidLoad() {
        super.viewDidLoad()
        
        objects.append(CakeItem(cakeTitle: "Item 1", cakeDescription: "This is a cake....", cakeImageUrl: nil))
        objects.append(CakeItem(cakeTitle: "Item 2", cakeDescription: "This is a cake....", cakeImageUrl: nil))
        objects.append(CakeItem(cakeTitle: "Item 2", cakeDescription: "This is a cake....", cakeImageUrl: nil))
        objects.append(CakeItem(cakeTitle: "Item 2", cakeDescription: "This is a cake....", cakeImageUrl: nil))
        objects.append(CakeItem(cakeTitle: "Item 2", cakeDescription: "This is a cake....", cakeImageUrl: nil))
    }

    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
    }

    // MARK: - Table View

    override func numberOfSections(in tableView: UITableView) -> Int {
        return 1
    }

    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return objects.count
    }

    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: "CakeCell", for: indexPath) as! CakeCell

        let cake = objects[indexPath.row]
        cell.titleLabel.text = cake.cakeTitle
        cell.descriptionLabel.text = cake.cakeDescription
        
        return cell
    }

    override func tableView(_ tableView: UITableView, canEditRowAt indexPath: IndexPath) -> Bool {
        return false
    }

    override func tableView(_ tableView: UITableView, commit editingStyle: UITableViewCell.EditingStyle, forRowAt indexPath: IndexPath) {
        if editingStyle == .delete {
            objects.remove(at: indexPath.row)
            tableView.deleteRows(at: [indexPath], with: .fade)
        } else if editingStyle == .insert {
            // Create a new instance of the appropriate class, insert it into the array, and add a new row to the table view.
        }
    }


}

